Restore `TRUSTED_PROXIES` behavior for Rails 5

micahlee opened the pull request:
### What does this PR do?
- _What's changed? Why were these changes made?_

This PR re-implements the `TRUSTED_PROXIES` Rack IP filter for Rails 5. This replaces the implementation previously provided by the `conjur-rack` gem for Rails 4.

This PR also adds rspec tests that verify the expected `request.ip` with the given scenarios for `REMOTE_ADDR`, `TRUSTED_PROXIES`, and `X-Forwarded-For`.

- _How should the reviewer approach this PR, especially if manual tests are required?_
- _Are there relevant screenshots you can add to the PR description?_

### What ticket does this PR close?
Connected to #1689 

### Checklists

#### Change log
- [x] The CHANGELOG has been updated, or
- [ ] This PR does not include user-facing changes and doesn't require a CHANGELOG update

#### Test coverage
- [x] This PR includes new unit and integration tests to go with the code changes, or
- [ ] The changes in this PR do not require tests

#### Documentation
- [x] Docs (e.g. `README`s) were updated in this PR, and/or there is a follow-on issue to update docs, or
- [ ] This PR does not require updating any documentation


#### TODO:
- [x] Cleanup fix code
- [x] Add tests for the `request.ip` behavior
- [x] Create issue/PR to remove monkey patch in `conjur-rack` for `trusted_proxy?` 
    Issue: https://github.com/cyberark/conjur-rack/issues/21
- [x] Update CHANGELOG

Tovli commented on spec/request/request_ip_spec.rb:
```diff
@@ -0,0 +1,163 @@
+      end
+    end
+
+    # The following spec exercise how variations on the X-Forwarded-For header
```
Hi,
I don't see a test that simulate our issue of having duplicate headers.
Currently we get from LB `X-Forwarded-For` but nginx is adding `X-Forwarded-For`
Need to either [fix this issue in nginx](https://github.com/conjurinc/appliance/pull/1300) or test for it so we can better simulate the production use

micahlee replied:
@Tovli, I am not trying to test for that yet, nor address that in this PR.

This PR is to resolve a regression found in Conjur from Rails 5 while work on the Client IP Sourcing doc here: https://github.com/cyberark/dap-wiki/pull/85.

I haven't gotten to the DAP portion in that document yet, and I expect this to come up then.

As a side note, I do agree that the issue should probably be fixed in DAP's nginx reverse proxy though.

Tovli commented on spec/request/request_ip_spec.rb:
```diff
@@ -0,0 +1,163 @@
+  let(:x_forwarded_for) { nil }
+
+  # --------------------------------------------------------------------
+  # Test Scenarios
```
I'm missing a test with CIDR range in the trusted proxies

micahlee replied:
Thanks, I've added tests to include trusted IP range examples using CIDR notation.

jonahx commented on lib/conjur/trusted_proxy_filter.rb:
```diff
@@ -0,0 +1,44 @@
+
+    def env_trusted_proxies
+      @env['TRUSTED_PROXIES']
+    end
```
I think I'd just inline this.

micahlee replied:
Yeah... I wish there was a better way to configure code climate for this. Inlining was my default, but code climate didn't like that "`@env['TRUSTED_PROXIES']` is called twice."

I'll inline again and add the exception comment.

micahlee commented:
@jonahx 

> I'd still like to move unit testing the object directly for the matrix tests, per our previous convo. Were you planning to do that in a different PR?

I'm torn on this. On one hand, I agree that the layers of unit tests would be the more proper way to do this. On the other hand, I do still have concerns about the fidelity of the tests, considering how mutable the ActiveDispatch::Request is as it passes through the Rack/Rails pipeline. I really do want the subject of these tests to be that request pipeline, rather than only the `TrustedProxiesFilter`. Or to assume too much about what the final state of `ActiveDispatch::Request` will be after the pipeline executes.

I _am_ willing to explore that refactor, but I do think that's better suited as a follow up PR. The ROI at this moment of that refactor seems low compared to the ROI of finishing the client IP sourcing document (which depends on this regression fix) on these points:

A) These tests do clearly communicate what the expected outcomes are for the inputs we care about.
B) These tests run fast now, so a refactor wouldn't greatly improve test cycle times.
C) We do get a high degree of assurance from these tests regarding the `request.ip` outcome, given the full Rack request pipeline is exercised. 

I created the issue here: https://github.com/cyberark/conjur/issues/1696. If there is an epic for the overall spec refactor you discussed during our team meeting Wednesday. I think this would be very appropriate in that epic.

> Though I really don't like the rspec convention whereby a function call with 3 parameters becomes 3 let statements followed by an include_examples to "call the function"

I agree with you, and I'd like to hear your suggestions for how to do it better.

* I considered inlining everything with something like:
    ```ruby
    request_env = { 'HTTP_AUTHORIZATION' => token_auth_header, 'REMOTE_ADDR' => '4.4.4.4' }
    headers = { 'X-Forwarded-For' => '3.3.3.3,5.5.5.2,5.5.5.3'
    ENV['TRUSTED_PROXIES'] = '4.4.4.4,5.5.5.0/24'
    get '/request_ip', env: request_env, headers: headers
    expect(body_hash['ip']).to eq('3.3.3.3')
    ```

  But this felt overly cumbersome, not very DRY, and made it hard to read the values that mattered.

* I considered writing my own helper method in vanilla Ruby to produce the result. Something akin to:
    ```ruby
    expect(request_ip(
      remote_addr: '4.4.4.4', 
      x_forwarded_for: '3.3.3.3,5.5.5.2,5.5.5.3',
      trusted_proxies: '4.4.4.4,5.5.5.0/24'
      )
    ).to eq('3.3.3.3')
    ```
  But this didn't feel remarkably different in outcome (or LOC) from the `let`s and `include_examples`, or in the spirit of `rspec`.

* Some of the `let`s are unnecessary since they are inherited from the containing context, so they could be removed. But I thought it was easier to read with each test's context clearly stated.

What do you think? Do any of those stand out as a clear improvement over the current code? Do you have any other suggestions I haven't considered yet?

alexkalish commented:
@micahlee: Are you planning on covering the use `TRUSTED_PROXIES` (including `evoke`, etc) in depth in your IP related `dap-wiki` PR?  If not, can you file a separate issue to do so?  I want ensure that we document this feature well in `dap-wiki` so that it can get into the docs for 11.7.  Thanks!

micahlee commented:
@alexkalish 
> Are you planning on covering the use TRUSTED_PROXIES (including evoke, etc) in depth in your IP related dap-wiki PR?

I'm not sure if this PR is where you want this discussion, rather than on the dap-wiki PR, but [the existing draft](https://github.com/cyberark/dap-wiki/pull/85) includes this as a reference discussion. [Jason's PR](https://github.com/cyberark/dap-wiki/pull/87) includes this in a tutorial/guide format (although I now see he's removed Conjur OSS from his doc). 

It would be helpful to know more specifically what gaps there are that need to be filled for allowing this to be officially documented. 

You can see my current "to do" list on[ the PR description here](https://github.com/cyberark/dap-wiki/pull/85). If Jason does not intend to cover OSS configuration of `TRUSTED_PROXIES` in his tutorial. If that is the case, I can create an issue for that.

alexkalish commented:
@micahlee: Actually, I forgot that @jvanderhoof was already working on this PR: https://github.com/cyberark/dap-wiki/pull/87.  Maybe that could serve to document this feature.  Could you take a look?

jonahx commented:
@micahlee 
- Your plan re: the refactor or lack thereof is ok with me.  It's good food for thought for future tests either way and eventually it would be nice to get standards out of it, but now is not the time for that.
- I prefer both of your alternatives to the `let` versions.  And I think "inherited" lets are especially problematic, and would always take repetition over that.
- I like the version with your bespoke helper most.  I think it not being "rspec"-ish is a non-issue.  In this case it seems to be a badge of honor.

micahlee commented:
@jonahx , this is still ready for your re-review when you're able. I added one more fixup commit today to re-used a shared method to create the test access token that I added in https://github.com/cyberark/conjur/pull/1698

I've left the fixups as separate commits to make the updates easier to review. Once you've had a chance to look at them, I'll squash them back down.

Thanks!

jonahx commented on spec/request/request_ip_spec.rb:
```diff
@@ -0,0 +1,150 @@
+  # By default "non-routable" IP addresses are trusted (according to this
+  # regular expression: https://github.com/rack/rack/blob/master/lib/rack/request.rb#L19)
+  #
+  # `127.0.0.1` is important as the address of the nginx proxy when used in DAP
```
Love these contextual comments.

jonahx commented on spec/request/request_ip_spec.rb:
```diff
@@ -0,0 +1,150 @@
+  end
+
+  # When X-Forwarded-For is set from an untrusted remote IP, we ignore it
+  it 'ignores the XFF header when the remote addr is untrusted' do
```
The comment here seems to just repeat the test description, so can be removed.

jonahx commented on spec/request/request_ip_spec.rb:
```diff
@@ -0,0 +1,150 @@
+  # If TRUSTED proxies are set, and are remote IP address is not trusted
+  # we expect to get the remote IP
+  it "doesn't trust the remote_addr if not included in TRUSTED_PROXIES" do
+    expect(
```
This comment too feels like repetition of the test case.
